apps/dash-3d-image-partitioning/app.py

```diff
@@ -0,0 +1,935 @@
+        # base64-encoded data between client and server: we let the browser
+        # handle how data from the browser can be written to the client's
+        # filesystem.
+        html.Div(
+            id="loader-wrapper",
+            children=[
+                # required so callback triggered by writing to "found-image-tensor-data"
+                # has an output
+                html.Div(id="dummy", style={"display": "none"}),
+                html.Div(id="dummy2", style={"display": "none"}, children=",0"),
+                # hidden elements so we can show/hide segmentations on 2d and 3d figures
+                html.Div(
+                    id="show-hide-seg-2d", children="show", style={"display": "none"}
+                ),
+                html.Div(
+                    id="show-hide-seg-3d", children="show", style={"display": "none"}
+                ),
+                dcc.Loading(
+                    id="graph-loading",
+                    type="circle",
+                    children=[
+                        html.A(id="download-link", download="found_image.nii",),
+                        # the image data of the found segmentation is stored
+                        # here before it is downloaded
+                        dcc.Store(id="found-image-tensor-data", data=""),
+                        html.Div(
+                            children=[
+                                html.Button(
+                                    "3D View",
+                                    id="view-select-button",
+                                    n_clicks=0,
+                                    style={"width": "25%"},
+                                ),
+                                html.Button(
+                                    "Hide Segmentation",
+                                    id="show-seg-check",
+                                    n_clicks=0,
+                                    style={"width": "25%"},
+                                ),
+                                html.Button(
+                                    "Download Selected Partitions",
+                                    id="download-button",
+                                    style={"width": "25%"},
+                                ),
+                                html.Button(
+                                    "Undo",
+                                    id="undo-button",
+                                    n_clicks=0,
+                                    style={"width": "12.5%"},
+                                ),
+                                html.Button(
+                                    "Redo",
+                                    id="redo-button",
+                                    n_clicks=0,
+                                    style={"width": "12.5%"},
+                                ),
+                            ],
+                            style={"display": "flex", "margin": "2px 0 2px 0"},
```

Comment:
is it useful to have css both in the app file and in specific files in assets?

Reply:
Yes and no, for instance, we could get away with writing all the layout in the `app.py` file. But for things like fonts it's easier to put them in `assets/style.css`. Personally I found it a little bit easier to put the CSS related to positioning in the `app.py` file because then it was possible to compute offsets, set variables at the top of the app, etc., which I found more intuitive than writing it in a separate CSS file. But I could move everything over to a separate file if it looks cleaner (some calculation is possible in pure CSS).